File: scenarios/identity-python-credential-chain/golden/credential_factory.py
from __future__ import annotations

import os

from azure.identity import (
    AzureCliCredential,
    AzurePowerShellCredential,
    ChainedTokenCredential,
    EnvironmentCredential,
    ManagedIdentityCredential,
    WorkloadIdentityCredential,
    aio,
)


def _managed_identity() -> ManagedIdentityCredential:
    return ManagedIdentityCredential(client_id=os.getenv("AZURE_CLIENT_ID"))


def _async_managed_identity() -> aio.ManagedIdentityCredential:
    return aio.ManagedIdentityCredential(client_id=os.getenv("AZURE_CLIENT_ID"))


def _dev_sync() -> ChainedTokenCredential:
    return ChainedTokenCredential(
        AzureCliCredential(),
        AzurePowerShellCredential(),
    )


def _ci_sync() -> ChainedTokenCredential:
    return ChainedTokenCredential(
        EnvironmentCredential(),
        WorkloadIdentityCredential(),
    )


def _production_sync() -> ChainedTokenCredential:
    return ChainedTokenCredential(
        _managed_identity(),
        WorkloadIdentityCredential(),
    )


def build_sync_credential(environment: str) -> ChainedTokenCredential:
    if environment == "dev":
        return _dev_sync()
    if environment == "ci":
        return _ci_sync()
    if environment == "production":
        return _production_sync()
    raise ValueError(f"Unknown environment: {environment}")


def _dev_async() -> aio.ChainedTokenCredential:
    return aio.ChainedTokenCredential(
        aio.AzureCliCredential(),
        aio.AzurePowerShellCredential(),
    )


def _ci_async() -> aio.ChainedTokenCredential:
    return aio.ChainedTokenCredential(
        aio.EnvironmentCredential(),
        aio.WorkloadIdentityCredential(),
    )


def _production_async() -> aio.ChainedTokenCredential:
    return aio.ChainedTokenCredential(
        _async_managed_identity(),
        aio.WorkloadIdentityCredential(),
    )


def build_async_credential(environment: str) -> aio.ChainedTokenCredential:
    if environment == "dev":
        return _dev_async()
    if environment == "ci":
        return _ci_async()
    if environment == "production":
        return _production_async()
    raise ValueError(f"Unknown environment: {environment}")


def strategy_for(environment: str) -> str:
    strategies = {
        "dev": "Azure CLI, then Azure PowerShell",
        "ci": "environment credential, then workload identity",
        "production": "managed identity, then workload identity",
    }
    return strategies[environment]

File: scenarios/identity-java-credential-chain/golden/src/main/java/com/example/CredentialFactory.java
package com.example;

import com.azure.core.credential.TokenCredential;
import com.azure.identity.AzureCliCredentialBuilder;
import com.azure.identity.AzurePowerShellCredentialBuilder;
import com.azure.identity.ChainedTokenCredentialBuilder;
import com.azure.identity.EnvironmentCredentialBuilder;
import com.azure.identity.ManagedIdentityCredentialBuilder;
import com.azure.identity.WorkloadIdentityCredentialBuilder;

public final class CredentialFactory {
    private CredentialFactory() {
    }

    public static TokenCredential buildCredential(
            DeploymentEnvironment environment) {
        return switch (environment) {
            case DEV -> createDevelopmentCredential();
            case CI -> createCiCredential();
            case PRODUCTION -> createProductionCredential();
        };
    }

    private static TokenCredential createDevelopmentCredential() {
        return new ChainedTokenCredentialBuilder()
                .addLast(new AzureCliCredentialBuilder().build())
                .addLast(new AzurePowerShellCredentialBuilder().build())
                .build();
    }

    private static TokenCredential createCiCredential() {
        return new ChainedTokenCredentialBuilder()
                .addLast(new EnvironmentCredentialBuilder().build())
                .addLast(new WorkloadIdentityCredentialBuilder().build())
                .build();
    }

    private static TokenCredential createProductionCredential() {
        return new ChainedTokenCredentialBuilder()
                .addLast(createManagedIdentityCredential())
                .addLast(new WorkloadIdentityCredentialBuilder().build())
                .build();
    }

    private static TokenCredential createManagedIdentityCredential() {
        String clientId = System.getenv("AZURE_CLIENT_ID");
        ManagedIdentityCredentialBuilder builder =
                new ManagedIdentityCredentialBuilder();
        if (clientId != null && !clientId.isBlank()) {
            builder.clientId(clientId);
        }
        return builder.build();
    }

    public static String strategyFor(DeploymentEnvironment environment) {
        return switch (environment) {
            case DEV -> "Azure CLI, then Azure PowerShell";
            case CI -> "environment credential, then workload identity";
            case PRODUCTION -> "managed identity, then workload identity";
        };
    }
}
